### ISSUE TITLE
UnboundLocalError: local variable 'err_code' referenced before assignment

### ISSUE THREAD

**Issue by shayanelhami** [python-pillow/Pillow#2071]
On this line 
[ImageFile.py Line 238](https://github.com/python-pillow/Pillow/blob/master/PIL/ImageFile.py#L238)

`if not self.map and not LOAD_TRUNCATED_IMAGES and err_code < 0:`

It happens when there is a ValueError exception on line 194:

``` python
    try:
        decoder.setimage(self.im, extents)
    except ValueError:
        continue
```

which leave no opportunity for err_code to be set.  
### What did you do?

Say I have a 128x128 png with one zip tile:
![128x128](https://cloud.githubusercontent.com/assets/1285365/17634055/9429cb88-60c8-11e6-91c2-0dac8d38a086.png)

and I use this code:

``` python
from PIL import Image
im = Image.open("128x128.png")
im.size = (32, 32)
im.save("32x32.png")
```
### What did you expect to happen?

Instead of an UnboundLocalError it should at least raise a meaningful error.
### What versions of Pillow and Python are you using?

Python 2.7.12, Pillow commit 533f4145

**Comment by homm** [comment]
We should make read-only property and raise on this:

```python
im.size = (32, 32)
```

**Comment by wiredfool** [comment]
Agreed.

**Comment by homm** [comment]
Actually, the size assigning is used at least [for icns files](https://github.com/python-pillow/Pillow/blob/master/Tests/test_file_icns.py#L47).

We'll need to create some smart property which will accept only allowed sizes and only before loading.

**Comment by radarhere** [comment]
@homm what do you mean by 'only allowed sizes'?

**Comment by PeterJCLaw** [comment]
Not sure if it's the same issue, but I've seen a similar symptom when trying to load a (corrupted) GIF file. The GIF in question causes a `ValueError` at the same location and then also the `UnboundLocalError`. My scenario didn't involve any assignment of the size of the image, just doing `Image.open("bad.gif").load()` was enough.

**Comment by alexwlchan** [comment]
@radarhere Rather than changing all the internal setters to use `size`, could you use `@size.setter` and implement the custom logic there? Something like:

```
@property
def size(self):
    return self._size

@size.setter
def size(self, value)
    # validation logic goes here...
    self._size = value
```

Then all the existing assignments continue to work without modifications, but we get the nice validation logic.

I’d offer a patch, but I confess that it’s not completely obvious to me what the conditions are that should trip those validation checks.

**Comment by radarhere** [comment]
@alexwlchan the problem is that the goal here is not to add validation logic, but to raise an error based on any external setting of the property. So then, in order to not raise an error wherever `size` is set internally, `_size` would need to be set directly by all internal Pillow code. That's perfectly possible, it's just not a small change.

**Comment by radarhere** [comment]
I have created PR #3203 as a suggestion for this.

**Comment by hugovk** [comment]
Should it give a deprecation warning to begin with, before raising in a later release? Or in this case is it okay to go straight to the raise?

**Comment by radarhere** [comment]
I've just found https://github.com/python-pillow/Pillow/issues/3044#issuecomment-375207867, where setting the size directly on a TIFF image is given as a solution.

**Comment by radarhere** [comment]
Resolved by #3203
